kube-runtime/src/scheduler.rs

```diff
@@ -16,11 +17,18 @@ use tokio_util::time::delay_queue::{self, DelayQueue};
 pub struct ScheduleRequest<T> {
     pub message: T,
     pub run_at: Instant,
+    // The reason behind the scheduling of this request. This field mainly
+    // exists to let us shadow the value of `controller::ReconcileRequest.reason`
+    // since the controller uses the scheduler to emit `controller::ReconcileRequest`s
+    // as a `message` and we can't update its `reason` field since message is bound to
+    // a generic parameter.
+    pub reason: String,
```

Comment:
Looks like this could be a `Option<ReconcileReason>`?

Reply:
yes, but having gone through the code, i realized that its heavily generic and makes no assumptions about its usage (it doesn't even import any controller modules), so i decided to keep it as a simple `String`.  see: https://github.com/kube-rs/kube/issues/1114#issuecomment-1619589804
i also experimented with `reason: S` where `S: IntoString` but that made the code unnecessarily complex.

Reply:
Thanks, I missed the comment in the issue!

With that explanation I would lean towards a generic reason and using `Option<T>` (where `T` is the reason), but should wait for others to chime in.

Reply:
The messages should be otherwise identical, so we could "just" define it to keep the sooner whole `ScheduleRequest.message`.

Reply:
One way to do that would be to add a `override_message: Option<T>` field to `ScheduledEntry`, and then emit that instead of the original message if it's `Some`.